ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/BatchValidator.scala

```diff
@@ -0,0 +1,148 @@
+// Copyright (c) 2020 The DAML Authors. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.validator
+
+import java.util.concurrent.{ArrayBlockingQueue, ThreadPoolExecutor, TimeUnit}
+
+import com.daml.ledger.participant.state.kvutils.DamlKvutils._
+import com.daml.ledger.participant.state.kvutils.Envelope
+import com.daml.ledger.participant.state.v1.ParticipantId
+import com.daml.ledger.validator.SubmissionValidator._
+import com.digitalasset.daml.lf.data.Time.Timestamp
+import com.digitalasset.daml.lf.engine.Engine
+import com.digitalasset.logging.ContextualizedLogger
+import com.google.common.util.concurrent.ThreadFactoryBuilder
+
+import scala.collection.JavaConverters._
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.Try
+
+// This is example code for validating a batch of submissions in parallel, including conflict detection.
+class BatchValidator[LogResult](
+    val allocateLogEntryId: () => DamlLogEntryId,
+    val ledgerStateAccess: LedgerStateAccess[LogResult]) {
+
+  // Queue for validation work.
+  val validationQueue = new ArrayBlockingQueue[Runnable](1024)
+  val nThreads = 2 * Runtime.getRuntime.availableProcessors()
+
+  // Executor that performs the validations.
+  // TODO(JM): We probably don't need anything this complicated. A fixed thread pool should be fine.
+  // Or if we want to keep things simpler we can just use an existing thread pool / execution context,
+  // though it would be useful for profiling purposes to have a dedicated thread pool.
+  val executor = new ThreadPoolExecutor(
```

Comment:
Nothing cleans this up. Can we make this `Closeable`, or better yet, injected?

Reply:
What do you mean injected?

Reply:
Passed in through the constructor. Basically, make it Somebody Else's Problem™.

Of course, you're also in charge of the Somebody Else. Making a `ResourceOwner` might help. See `DbDispatcher` for an example.